File: frontend/js/views/demand_for_payment.js
// Demand for payment (demand letter) generator — totals principal + interest +
// fees and computes the pay-by date, via /calc/demand-for-payment. Previews
// live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderDemandForPayment(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.demand.h1.title">// DEMAND FOR PAYMENT</span></h1>
        <p class="muted small" data-i18n="view.demand.hint.intro">
            The formal written demand a creditor sends before pursuing collection or suit — often a
            prerequisite to small-claims or collection action. It totals the principal, accrued interest,
            and late fees, and computes the pay-by date from the demand date plus the response window.
            Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.demand.h2.inputs">Letter details</h2>
            <form id="demand-form" class="inline-form">
                <label><span data-i18n="view.demand.label.state">Governing state</span>
                    <input type="text" name="governing_state" value="New York" required></label>
                <label><span data-i18n="view.demand.label.creditor">Creditor name</span>
                    <input type="text" name="creditor_name" value=""></label>
                <label><span data-i18n="view.demand.label.creditor_address">Creditor address</span>
                    <input type="text" name="creditor_address" value=""></label>
                <label><span data-i18n="view.demand.label.creditor_phone">Creditor phone</span>
                    <input type="text" name="creditor_phone" value=""></label>
                <label><span data-i18n="view.demand.label.debtor">Debtor name</span>
                    <input type="text" name="debtor_name" value=""></label>
                <label><span data-i18n="view.demand.label.description">What is owed for</span>
                    <input type="text" name="debt_description" value="Invoice #1042, unpaid"></label>
                <label><span data-i18n="view.demand.label.principal">Principal ($)</span>
                    <input type="number" step="0.01" min="0" name="principal_usd" value="5000" required></label>
                <label><span data-i18n="view.demand.label.interest">Accrued interest ($)</span>
                    <input type="number" step="0.01" min="0" name="accrued_interest_usd" value="150"></label>
                <label><span data-i18n="view.demand.label.fees">Late fees ($)</span>
                    <input type="number" step="0.01" min="0" name="late_fees_usd" value="50"></label>
                <label><span data-i18n="view.demand.label.demand_date">Demand date</span>
                    <input type="date" name="demand_date" value="2026-06-01" required></label>
                <label><span data-i18n="view.demand.label.response_days">Response window (days)</span>
                    <input type="number" step="1" min="1" name="response_days" value="15" required></label>
                <label><span data-i18n="view.demand.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.demand.ph.statute'))}"></label>
            </form>
        </div>
        <div id="demand-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#demand-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            creditor_name: (fd.get('creditor_name') || '').trim(),
            creditor_address: (fd.get('creditor_address') || '').trim(),
            creditor_phone: (fd.get('creditor_phone') || '').trim(),
            debtor_name: (fd.get('debtor_name') || '').trim(),
            principal_usd: Number(fd.get('principal_usd')) || 0,
            accrued_interest_usd: Number(fd.get('accrued_interest_usd')) || 0,
            late_fees_usd: Number(fd.get('late_fees_usd')) || 0,
            debt_description: (fd.get('debt_description') || '').trim(),
            demand_date: fd.get('demand_date'),
            response_days: Math.round(Number(fd.get('response_days')) || 0),
            governing_state: (fd.get('governing_state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcDemandForPayment(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.demand.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#demand-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card neg"><div class="label" data-i18n="view.demand.card.total">Total now due</div>
                    <div class="value">${money(doc.total_due_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.demand.card.principal">Principal</div>
                    <div class="value">${money(doc.principal_usd)}</div></div>
                <div class="card neg"><div class="label" data-i18n="view.demand.card.payby">Pay by</div>
                    <div class="value">${esc(doc.pay_by_date || '—')}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="demand-copy" type="button" data-i18n="view.demand.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="demand-download" type="button" data-i18n="view.demand.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#demand-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.demand.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.demand.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#demand-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'demand-for-payment.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/demand_for_payment.rs
//! Demand for payment (demand letter) — the formal written demand a creditor
//! sends before pursuing collection or suit. It totals the principal, accrued
//! interest, and late fees, computes the pay-by date from the demand date plus
//! the response window, and assembles the letter. Often a prerequisite to small-
//! claims or collection action. Drafting aid, not legal advice.

use chrono::{Duration, NaiveDate};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct DemandInput {
    pub creditor_name: String,
    pub creditor_address: String,
    #[serde(default)]
    pub creditor_phone: String,
    pub debtor_name: String,
    pub principal_usd: f64,
    #[serde(default)]
    pub accrued_interest_usd: f64,
    #[serde(default)]
    pub late_fees_usd: f64,
    /// What the debt is for (invoice #, goods, services, loan…).
    pub debt_description: String,
    /// Date the demand is sent (YYYY-MM-DD).
    pub demand_date: String,
    /// Days the debtor has to pay before further action.
    pub response_days: i64,
    pub governing_state: String,
    #[serde(default)]
    pub statute_citation: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DemandLetter {
    pub title: String,
    pub principal_usd: f64,
    pub accrued_interest_usd: f64,
    pub late_fees_usd: f64,
    pub total_due_usd: f64,
    pub pay_by_date: String,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &DemandInput) -> DemandLetter {
    let total = i.principal_usd + i.accrued_interest_usd + i.late_fees_usd;

    let pay_by = NaiveDate::parse_from_str(&i.demand_date, "%Y-%m-%d")
        .map(|d| (d + Duration::days(i.response_days)).format("%Y-%m-%d").to_string())
        .unwrap_or_default();

    let citation = i.statute_citation.trim();
    let pursuant = if citation.is_empty() {
        format!(
            "This demand and any subsequent action are made under the laws of the State of {}.",
            i.governing_state
        )
    } else {
        format!(
            "This demand and any subsequent action are made under the laws of the State of {} ({}).",
            i.governing_state, citation
        )
    };

    // Build the itemized statement; only show interest / fees lines when nonzero.
    let mut lines = vec![
        format!("Amount owed for: {}", i.debt_description),
        format!("  Principal: {}", money(i.principal_usd)),
    ];
    if i.accrued_interest_usd != 0.0 {
        lines.push(format!("  Accrued interest: {}", money(i.accrued_interest_usd)));
    }
    if i.late_fees_usd != 0.0 {
        lines.push(format!("  Late fees: {}", money(i.late_fees_usd)));
    }
    lines.push(format!("  Total now due: {}", money(total)));

    let clauses = vec![
        DocClause { heading: "To".into(), body: i.debtor_name.clone() },
        DocClause {
            heading: "1. Statement of Debt".into(),
            body: lines.join("\n"),
        },
        DocClause {
            heading: "2. Demand for Payment".into(),
            body: format!(
                "Demand is hereby made for payment of the total amount of {} on or before {} ({} days from the date of this letter). Payment should be made to {} at {}.",
                money(total),
                pay_by,
                i.response_days,
                i.creditor_name,
                i.creditor_address
            ),
        },
        DocClause {
            heading: "3. Consequences of Non-Payment".into(),
            body: "If payment is not received by the date stated above, the undersigned may pursue all available remedies, which may include referral to a collection agency, reporting of the debt, and the commencement of legal proceedings to recover the amount due together with any interest, costs, and fees allowed by law. This letter is an attempt to collect a debt.".into(),
        },
        DocClause { heading: "4. Governing Law".into(), body: pursuant },
        DocClause {
            heading: "Signature".into(),
            body: format!(
                "Signature: ____________________  Date: {}\n{}\n{}{}",
                i.demand_date,
                i.creditor_name,
                i.creditor_address,
                if i.creditor_phone.is_empty() {
                    String::new()
                } else {
                    format!("\nTelephone: {}", i.creditor_phone)
                }
            ),
        },
    ];

    DemandLetter {
        title: "Demand for Payment".into(),
        principal_usd: i.principal_usd,
        accrued_interest_usd: i.accrued_interest_usd,
        late_fees_usd: i.late_fees_usd,
        total_due_usd: total,
        pay_by_date: pay_by,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-6
    }

    fn base() -> DemandInput {
        DemandInput {
            creditor_name: "Acme LLC".into(),
            creditor_address: "1 Commerce St".into(),
            creditor_phone: String::new(),
            debtor_name: "John Debtor".into(),
            principal_usd: 5000.0,
            accrued_interest_usd: 150.0,
            late_fees_usd: 50.0,
            debt_description: "Invoice #1042, unpaid".into(),
            demand_date: "2026-06-01".into(),
            response_days: 15,
            governing_state: "New York".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn total_sums_principal_interest_fees() {
        assert!(close(generate(&base()).total_due_usd, 5200.0));
    }

    #[test]
    fn pay_by_is_demand_plus_response() {
        // 2026-06-01 + 15 days = 2026-06-16.
        assert_eq!(generate(&base()).pay_by_date, "2026-06-16");
    }

    #[test]
    fn zero_interest_and_fees_total_is_principal() {
        let d = generate(&DemandInput {
            accrued_interest_usd: 0.0,
            late_fees_usd: 0.0,
            ..base()
        });
        assert!(close(d.total_due_usd, 5000.0));
        // Interest/fees lines omitted from the statement when zero.
        let stmt = d.clauses.iter().find(|c| c.heading.contains("Statement")).unwrap();
        assert!(!stmt.body.contains("Accrued interest"));
        assert!(!stmt.body.contains("Late fees"));
    }

    #[test]
    fn statement_lists_components_when_present() {
        let stmt = generate(&base())
            .clauses
            .into_iter()
            .find(|c| c.heading.contains("Statement"))
            .unwrap();
        assert!(stmt.body.contains("Accrued interest: $150.00"));
        assert!(stmt.body.contains("Late fees: $50.00"));
        assert!(stmt.body.contains("Total now due: $5200.00"));
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&DemandInput {
            statute_citation: "FDCPA 15 U.S.C. § 1692".into(),
            ..base()
        });
        assert_eq!(d.statutory_citation, "FDCPA 15 U.S.C. § 1692");
        assert!(d.clauses.iter().any(|c| c.body.contains("FDCPA 15 U.S.C. § 1692")));
    }

    #[test]
    fn bad_date_yields_empty_pay_by() {
        let d = generate(&DemandInput { demand_date: "x".into(), ..base() });
        assert_eq!(d.pay_by_date, "");
    }
}
